crates/leon/src/macros.rs

```diff
@@ -48,3 +48,140 @@ macro_rules! template {
         }, ::core::option::Option::Some($default))
     };
 }
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __template2_item {
+    () => {};
+    ({ $key:literal }) => {
+        $crate::Item::Key($key)
+    };
+    ( $text:literal ) => {
+        $crate::Item::Text($text)
+    };
+}
+
+/// Construct a template constant using syntax similar to the template to be
+/// passed to [`Template::parse`].
+///
+/// This is essentially a shorthand for:
+///
+/// ```
+/// use leon::{Item, Template};
+/// Template::new({
+///     const ITEMS: &'static [Item<'static>] = &[Item::Text("Hello "), Item::Key("name")];
+///     ITEMS
+/// }, Some("world"));
+/// ```
+///
+/// # Examples
+///
+/// ```
+/// assert_eq!(
+///     leon::template2!("Hello ", {"name"})
+///         .render(&[("name", "Магда Нахман")])
+///         .unwrap(),
+///     "Hello Магда Нахман",
+/// );
+/// ```
+#[macro_export]
+macro_rules! template2 {
```

Comment:
That's nicer, let's just replace the original one with it

Reply:
But removing the original will be a breaking change that requires v2.0.0

Reply:
Yeah, that's fine.

Reply:
I will remove that once I fix the strange lifetime error.

BTW, do you have any idea how to fix that?
I think it could be a compiler bug, though I read somewhere that the borrow checker can reject valid code since it's better to have false positive than false negative.

Reply:
My idea is, uh, "excessive use of lifetimes, clone more".

But more seriously maybe there's a future involved that implies a static lifetime for some of its arguments and that's poisoning everything

Reply:
> My idea is, uh, "excessive use of lifetimes, clone more".

Yeah that also get mentioned in the reddit thread, perhaps I should remove some of the references.

But not sure how to remove reference to `self.target_data.meta.pkg_url.as_deref()` though, because to remove that reference we have to make `pkg_url` a self-reference object, but then the final `pkg_urls` should be an `Either` implementing `Iterator<Item = Template<'_>>`.

> But more seriously maybe there's a future involved that implies a static lifetime for some of its arguments and that's poisoning everything

Yeah it definitely could be, it's caused by [commit c2ecfd4e62501c1e0f0ec917f90947dc7429f1ed](https://github.com/cargo-bins/cargo-binstall/pull/933/commits/c2ecfd4e62501c1e0f0ec917f90947dc7429f1ed) where I started to return `Template<'static>` instead of `String` in `RepositoryHost::get_default_pkg_url_template`:

```rust
impl RepositoryHost {
    pub fn get_default_pkg_url_template(
        self,
    ) -> Option<impl Iterator<Item = Template<'static>> + Clone + 'static>;
}
```

and then I also move the `Template::parse` earlier before the loop, right after call to `RepositoryHost::resolve_final_repo_url`.

Also removed one argument from `GhCrateMeta::launch_baseline_find_tasks`, which could be the suspect for causing this issue since it really has a lot of lifetime used:

```rust
type FindTaskRes = Result<Option<(Url, PkgFmt)>, BinstallError>;

impl GhCrateMeta {
    fn launch_baseline_find_tasks<'a>(
        &'a self,
        pkg_fmt: PkgFmt,
        pkg_url: &'a Template<'a>,
        repo: Option<&'a str>,
    ) -> impl Iterator<Item = impl Future<Output = FindTaskRes> + 'static> + 'a;
```

Reply:
I've replaced `leon::template!` with the new impl.